credentials/tls/certprovider/distributor_test.go

```diff
@@ -44,105 +45,113 @@ func (s) TestDistributor(t *testing.T) {
+	if gotKM, err := dist.KeyMaterial(ctx); gotKM != nil || !errors.Is(err, errProviderTestInternal) {
+		t.Fatalf("KeyMaterial() = {%v, %v}, want {nil, %v}", gotKM, err, errProviderTestInternal)
+	}
 
-	waitAndDo(t, proceedCh, errCh, func() {
-		dist.Set(&wantKM2, nil)
-	})
+	// Stop the distributor and  KeyMaterial() should return errProviderClosed.
+	dist.Stop()
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	if km, err := dist.KeyMaterial(ctx); !errors.Is(err, errProviderClosed) {
+		t.Fatalf("KeyMaterial() = {%v, %v}, want {nil, %v}", km, err, errProviderClosed)
+	}
+}
 
-	waitAndDo(t, proceedCh, errCh, func() {
-		dist.Set(&wantKM2, errProviderTestInternal)
-	})
+// TestDistributorConcurrency invokes methods on the distributor in parallel.
+func (s) TestDistributorConcurrency(t *testing.T) {
+	dist := NewDistributor()
 
-	waitAndDo(t, proceedCh, errCh, func() {
-		dist.Stop()
-	})
-}
+	// Read cert/key files from testdata.
+	km1 := loadKeyMaterials(t, "x509/server1_cert.pem", "x509/server1_key.pem", "x509/client_ca_cert.pem")
+	km2 := loadKeyMaterials(t, "x509/server2_cert.pem", "x509/server2_key.pem", "x509/client_ca_cert.pem")
 
-func waitAndDo(t *testing.T, proceedCh chan struct{}, errCh chan error, do func()) {
-	t.Helper()
+	var wg sync.WaitGroup
+	errCh := make(chan error, 1)
+
+	// Push key material into the distributor from here and spawn a goroutine to
+	// verify that the distributor returns the expected keyMaterial.
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	wg.Add(1)
+	go func() {
+		waitForKeyMaterial(ctx, dist, km1, nil, errCh)
+		wg.Done()
+	}()
+	dist.Set(km1, nil)
+	wg.Wait()
+	if err := <-errCh; err != nil {
+		t.Fatal(err)
+	}
 
-	timer := time.NewTimer(2 * defaultTestTimeout)
-	select {
-	case <-timer.C:
-		t.Fatalf("test timed out when waiting for event from distributor")
-	case <-proceedCh:
-		do()
-	case err := <-errCh:
+	// Push new key material into the distributor from here and spawn a
+	// goroutine to verify that the distributor returns the expected keyMaterial
+	// eventually.
+	ctx, cancel = context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+	wg.Add(1)
+	go func() {
+		waitForKeyMaterial(ctx, dist, km2, nil, errCh)
+		wg.Done()
+	}()
+	dist.Set(km2, nil)
+	wg.Wait()
+	if err := <-errCh; err != nil {
 		t.Fatal(err)
 	}
 }
+
+// waitForKeyMaterial reads key material from the given distributor until one of
+// the following conditions are met:
+// 1. Returned keyMaterial and error matches wantKM and wantErr.
+// 2. Provider ctx deadline expires.
+func waitForKeyMaterial(ctx context.Context, dist *Distributor, wantKM *KeyMaterial, wantErr error, errCh chan error) {
+	var (
+		gotKM *KeyMaterial
+		err   error
+	)
+
+	for {
+		if gotErr := readAndVerifyKeyMaterial(ctx, dist, wantKM); gotErr == wantErr {
+			errCh <- nil
+			return
+
+		}
+		if errors.Is(err, context.DeadlineExceeded) {
```

Comment:
`if err != nil` ?

Reply:
No, I only want to exit on context deadline errors here. Basically, distributor could have `km1` when this function is called. So, the first call to `readAndVerifyKeyMaterial` will fail. Checking only for context deadline errors here will make sure that we retry till we eventually get what key material we are looking for (or the context expires, at which point, we are done).